// @ionic/cli-framework/lib/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_fs_1 = require("@ionic/utils-fs");
const fs = require("fs");
const lodash = require("lodash");
const path = require("path");
const writeFileAtomic = require("write-file-atomic");
class BaseConfig {
    constructor(p, { pathPrefix = [] } = {}) {
        this.p = p;
        this.pathPrefix = pathPrefix;
    }
    get file() {
        const contents = fs.readFileSync(this.p, 'utf8');
        return JSON.parse(contents);
    }
    get c() {
        try {
            const file = this.file;
            const navigated = this.pathPrefix.length === 0 ? file : lodash.get(file, [...this.pathPrefix]);
            const config = typeof navigated === 'object' ? navigated : {};
            return lodash.assign({}, this.provideDefaults(config), config);
        }
        catch (e) {
            if (e.code === 'ENOENT') {
                return this.provideDefaults({});
            }
            if (e.name === 'SyntaxError') {
                writeFileAtomic.sync(this.p, '');
                return this.provideDefaults({});
            }
            throw e;
        }
    }
    set c(value) {
        const v = this.pathPrefix.length === 0 ? value : lodash.set(this.file, [...this.pathPrefix], value);
        utils_fs_1.mkdirpSync(path.dirname(this.p));
        writeFileAtomic.sync(this.p, JSON.stringify(v, undefined, 2));
    }
    get(property, defaultValue) {
        const value = this.c[property];
        if (defaultValue && typeof value === 'undefined') {
            return defaultValue;
        }
        return value;
    }
    set(property, value) {
        const config = this.c;
        config[property] = value;
        this.c = config;
    }
    unset(property) {
        const config = this.c;
        delete config[property];
        this.c = config;
    }
}
exports.BaseConfig = BaseConfig;
