// @ionic/cli-framework/lib/help.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const lodash = require("lodash");
const guards_1 = require("../guards");
const array_1 = require("../utils/array");
const format_1 = require("../utils/format");
const colors_1 = require("./colors");
const validators_1 = require("./validators");
const DEFAULT_DOTS_WIDTH = 25;
class HelpFormatter {
    constructor({ colors }) {
        this.colors = colors ? colors : colors_1.DEFAULT_COLORS;
    }
}
exports.HelpFormatter = HelpFormatter;
class NamespaceHelpFormatter extends HelpFormatter {
    constructor({ location, namespace, colors }) {
        super({ colors });
        this.dotswidth = DEFAULT_DOTS_WIDTH;
        this.location = location;
        this.namespace = namespace;
    }
    getNamespaceMetadata() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!this._metadata) {
                this._metadata = yield this.namespace.getMetadata();
            }
            return this._metadata;
        });
    }
    getNamespaceFullName() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!this._fullName) {
                this._fullName = this.location.path.map(([p]) => p).join(' ');
            }
            return this._fullName;
        });
    }
}
exports.NamespaceHelpFormatter = NamespaceHelpFormatter;
class NamespaceStringHelpFormatter extends NamespaceHelpFormatter {
    formatHeader() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { strong, input } = this.colors;
            const fullName = yield this.getNamespaceFullName();
            const summary = yield this.formatSummary();
            const description = yield this.formatDescription();
            return (`\n  ${strong(`${input(fullName)}${summary}`)}` +
                (description ? `\n\n    ${description}` : '') + '\n');
        });
    }
    formatSummary() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const fullName = yield this.getNamespaceFullName();
            const metadata = yield this.getNamespaceMetadata();
            const summary = ((this.formatBeforeSummary ? yield this.formatBeforeSummary() : '') +
                metadata.summary +
                (this.formatAfterSummary ? yield this.formatAfterSummary() : ''));
            const wrappedSummary = format_1.wordWrap(summary, { indentation: fullName.length + 5 });
            return wrappedSummary ? ` - ${wrappedSummary}` : '';
        });
    }
    formatDescription() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const metadata = yield this.getNamespaceMetadata();
            if (!metadata.description) {
                return '';
            }
            return format_1.wordWrap(metadata.description.trim(), { indentation: 4 });
        });
    }
    getExtraOptions() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return [];
        });
    }
    formatUsage() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { strong, weak, input } = this.colors;
            const fullName = yield this.getNamespaceFullName();
            const options = ['--help', ...(yield this.getExtraOptions())];
            const usageLines = [
                `<command> ${weak('[<args>]')} ${options.map(opt => weak('[' + opt + ']')).join(' ')} ${weak('[options]')}`,
            ];
            return (`\n  ${strong('Usage')}:` +
                `\n\n    ${usageLines.map(u => `${weak('$')} ${input(fullName + ' ' + u)}`).join('\n    ')}\n`);
        });
    }
    formatCommands() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const commands = yield this.namespace.getCommandMetadataList();
            return this.formatCommandGroup('Commands', commands);
        });
    }
    formatCommandGroup(title, commands) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { strong } = this.colors;
            const filterCallback = this.filterCommandCallback;
            const filteredCommands = filterCallback ? yield array_1.filter(commands, (cmd) => tslib_1.__awaiter(this, void 0, void 0, function* () { return filterCallback(cmd); })) : commands;
            const [cmdDetails, nsDetails] = yield Promise.all([
                this.getListOfCommandDetails(filteredCommands.filter(cmd => cmd.namespace === this.namespace)),
                this.getListOfNamespaceDetails(filteredCommands.filter(cmd => cmd.namespace !== this.namespace)),
            ]);
            const details = [...cmdDetails, ...nsDetails];
            if (details.length === 0) {
                return '';
            }
            details.sort();
            return (`\n  ${strong(title)}:` +
                `\n\n    ${details.join('\n    ')}\n`);
        });
    }
    getListOfCommandDetails(commands) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { weak, input } = this.colors;
            const fullCmd = commands.map(cmd => lodash.tail(cmd.path).map(([p]) => p).join(' '));
            const fillStringArray = format_1.generateFillSpaceStringList(fullCmd, this.dotswidth, weak('.'));
            const formattedCommands = yield Promise.all(commands.map((cmd, index) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                const summary = ((this.formatBeforeCommandSummary ? yield this.formatBeforeCommandSummary(cmd) : '') +
                    cmd.summary +
                    (yield this.formatAfterCommandSummary(cmd)));
                const wrappedSummary = format_1.wordWrap(summary, { indentation: this.dotswidth + 6 });
                return `${input(lodash.tail(cmd.path).map(([p]) => p).join(' '))}${wrappedSummary ? ' ' + fillStringArray[index] + ' ' + wrappedSummary : ''}`;
            })));
            return formattedCommands;
        });
    }
    getListOfNamespaceDetails(commands) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { weak, input } = this.colors;
            const namespaces = yield this.namespace.groupCommandsByNamespace(commands);
            const fillStringArray = format_1.generateFillSpaceStringList(namespaces.map(({ name }) => name + ' <subcommand>'), this.dotswidth, weak('.'));
            const formattedNamespaces = yield Promise.all(namespaces.map((meta, i) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                const summary = ((this.formatBeforeNamespaceSummary ? yield this.formatBeforeNamespaceSummary(meta, meta.commands) : '') +
                    meta.summary +
                    (yield this.formatAfterNamespaceSummary(meta, meta.commands)));
                const wrappedSummary = format_1.wordWrap(summary, { indentation: this.dotswidth + 6 });
                return `${input(meta.name + ' <subcommand>')}${wrappedSummary ? ' ' + fillStringArray[i] + ' ' + wrappedSummary : ''}`;
            })));
            return formattedNamespaces;
        });
    }
    /**
     * Insert text that appears after a commands's summary.
     *
     * @param meta: The metadata of the command.
     */
    formatAfterCommandSummary(meta) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { weak, input } = this.colors;
            const aliases = meta.aliases.map(alias => lodash.tail(alias.split(' ')).join(' '));
            const formattedAliases = aliases.length > 0 ? weak('(alias' + (aliases.length === 1 ? '' : 'es') + ': ') + aliases.map(a => input(a)).join(', ') + weak(')') : '';
            return formattedAliases ? ` ${formattedAliases}` : '';
        });
    }
    /**
     * Insert text that appears after a namespace's summary.
     *
     * @param meta The metadata of the namespace.
     * @param commands An array of the metadata of the namespace's commands.
     */
    formatAfterNamespaceSummary(meta, commands) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { weak, input } = this.colors;
            const formattedSubcommands = commands.length > 0 ? `${weak('(subcommands:')} ${commands.map(c => input(c.name)).join(', ')}${weak(')')}` : '';
            const formattedAliases = meta.aliases.length > 0 ? `${weak('(alias' + (meta.aliases.length === 1 ? '' : 'es') + ': ') + meta.aliases.map(a => input(a)).join(', ') + weak(')')}` : '';
            return `${formattedSubcommands ? ` ${formattedSubcommands}` : ''}${formattedAliases ? ` ${formattedAliases}` : ''}`;
        });
    }
    format() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return ((yield this.formatHeader()) +
                (yield this.formatUsage()) +
                (yield this.formatCommands()) +
                '\n');
        });
    }
}
exports.NamespaceStringHelpFormatter = NamespaceStringHelpFormatter;
class CommandHelpFormatter extends HelpFormatter {
    constructor({ location, command, metadata, colors }) {
        super({ colors });
        this.dotswidth = DEFAULT_DOTS_WIDTH;
        this.location = location;
        this.command = command;
        this._hydratedMetadata = metadata;
    }
    getCommandMetadata() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (this._hydratedMetadata) {
                return this._hydratedMetadata;
            }
            if (!this._metadata) {
                this._metadata = yield this.command.getMetadata({ location: this.location });
            }
            return this._metadata;
        });
    }
    getCommandFullName() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!this._fullName) {
                this._fullName = this.location.path.map(([p]) => p).join(' ');
            }
            return this._fullName;
        });
    }
}
exports.CommandHelpFormatter = CommandHelpFormatter;
class CommandStringHelpFormatter extends CommandHelpFormatter {
    formatHeader() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { strong, input } = this.colors;
            const fullName = yield this.getCommandFullName();
            const summary = yield this.formatSummary();
            const description = yield this.formatDescription();
            return (`\n  ${strong(`${input(fullName)}${summary}`)}` +
                (description ? `\n\n    ${description}` : '') + '\n');
        });
    }
    formatSummary() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const fullName = yield this.getCommandFullName();
            const metadata = yield this.getCommandMetadata();
            const summary = ((this.formatBeforeSummary ? yield this.formatBeforeSummary() : '') +
                metadata.summary +
                (this.formatAfterSummary ? yield this.formatAfterSummary() : ''));
            const wrappedSummary = format_1.wordWrap(summary, { indentation: fullName.length + 5 });
            return wrappedSummary ? ` - ${wrappedSummary}` : '';
        });
    }
    formatDescription() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const metadata = yield this.getCommandMetadata();
            if (!metadata.description) {
                return '';
            }
            return format_1.wordWrap(metadata.description.trim(), { indentation: 4 });
        });
    }
    formatInlineInput(input) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (input.validators && input.validators.includes(validators_1.validators.required)) {
                return '<' + input.name + '>';
            }
            return '[<' + input.name + '>]';
        });
    }
    formatUsage() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { strong, weak, input } = this.colors;
            const fullName = yield this.getCommandFullName();
            const metadata = yield this.getCommandMetadata();
            const filterCallback = this.filterOptionCallback;
            const options = metadata.options ? metadata.options : [];
            const filteredOptions = filterCallback ? yield array_1.filter(options, (opt) => tslib_1.__awaiter(this, void 0, void 0, function* () { return filterCallback(opt); })) : options;
            const formattedInputs = metadata.inputs ? yield Promise.all(metadata.inputs.map((i) => tslib_1.__awaiter(this, void 0, void 0, function* () { return this.formatInlineInput(i); }))) : [];
            return (`\n  ${strong('Usage')}:` +
                `\n\n    ${weak('$')} ${input(fullName + (formattedInputs ? ' ' + formattedInputs.join(' ') : ''))}${filteredOptions.length > 0 ? ' ' + input('[options]') : ''}\n`);
        });
    }
    formatInputs() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { strong, weak, input } = this.colors;
            const metadata = yield this.getCommandMetadata();
            const inputs = metadata.inputs ? metadata.inputs : [];
            if (inputs.length === 0) {
                return '';
            }
            const fillStrings = format_1.generateFillSpaceStringList(inputs.map(i => i.name), this.dotswidth, weak('.'));
            const inputLineFn = ({ name, summary }, index) => {
                const optionList = input(`${name}`);
                const wrappedSummary = format_1.wordWrap(summary, { indentation: this.dotswidth + 6 });
                return `${optionList} ${fillStrings[index]} ${wrappedSummary}`;
            };
            return (`\n  ${strong('Inputs')}:` +
                `\n\n    ${inputs.map(inputLineFn).join('\n    ')}\n`);
        });
    }
    formatOptionLine(opt) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { weak } = this.colors;
            const optionName = formatOptionName(opt, { colors: this.colors });
            const optionNameLength = format_1.stringWidth(optionName);
            const fullLength = optionNameLength > this.dotswidth ? optionNameLength + 1 : this.dotswidth;
            const fullDescription = ((this.formatBeforeOptionSummary ? yield this.formatBeforeOptionSummary(opt) : '') +
                opt.summary +
                (yield this.formatAfterOptionSummary(opt)));
            const wrappedDescription = format_1.wordWrap(fullDescription, { indentation: this.dotswidth + 6 });
            return `${optionName} ${weak('.').repeat(fullLength - optionNameLength)} ${wrappedDescription}`;
        });
    }
    formatAfterOptionSummary(opt) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return this.formatOptionDefault(opt);
        });
    }
    formatOptionDefault(opt) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { weak, input } = this.colors;
            if (typeof opt.default === 'string') {
                return weak(' (default: ') + input(opt.default) + weak(')');
            }
            else {
                return '';
            }
        });
    }
    formatOptions() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const metadata = yield this.getCommandMetadata();
            const options = metadata.options ? metadata.options : [];
            return this.formatOptionsGroup('Options', options);
        });
    }
    formatOptionsGroup(title, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { strong } = this.colors;
            const filterCallback = this.filterOptionCallback;
            const filteredOptions = filterCallback ? yield array_1.filter(options, (opt) => tslib_1.__awaiter(this, void 0, void 0, function* () { return filterCallback(opt); })) : options;
            if (filteredOptions.length === 0) {
                return '';
            }
            const formattedOptions = yield Promise.all(filteredOptions.map((option) => tslib_1.__awaiter(this, void 0, void 0, function* () { return this.formatOptionLine(option); })));
            return (`\n  ${strong(title)}:` +
                `\n\n    ${formattedOptions.join('\n    ')}\n`);
        });
    }
    formatExamples() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { strong, weak, input } = this.colors;
            const metadata = yield this.getCommandMetadata();
            const fullName = yield this.getCommandFullName();
            if (!metadata.exampleCommands || !Array.isArray(metadata.exampleCommands)) {
                return '';
            }
            const exampleLines = metadata.exampleCommands.map(cmd => {
                const sepIndex = cmd.indexOf('-- ');
                cmd = sepIndex === -1 ? input(cmd) : input(cmd.substring(0, sepIndex)) + cmd.substring(sepIndex);
                const wrappedCmd = format_1.wordWrap(cmd, { indentation: 12, append: ' \\' });
                return `${weak('$')} ${input(fullName)}${wrappedCmd ? ' ' + wrappedCmd : ''}`;
            });
            return (`\n  ${strong('Examples')}:` +
                `\n\n    ${exampleLines.join('\n    ')}\n`);
        });
    }
    format() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return ((yield this.formatHeader()) +
                (yield this.formatUsage()) +
                (yield this.formatInputs()) +
                (yield this.formatOptions()) +
                (yield this.formatExamples()) +
                '\n');
        });
    }
}
exports.CommandStringHelpFormatter = CommandStringHelpFormatter;
class NamespaceSchemaHelpFormatter extends NamespaceHelpFormatter {
    format() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return JSON.stringify(yield this.serialize());
        });
    }
    serialize() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const metadata = yield this.getNamespaceMetadata();
            const commands = yield this.namespace.getCommandMetadataList();
            return {
                name: metadata.name,
                summary: metadata.summary,
                description: metadata.description ? metadata.description : '',
                groups: metadata.groups ? metadata.groups : [],
                commands: yield this.formatCommandGroup(commands),
                aliases: [],
            };
        });
    }
    formatCommandGroup(commands) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const filterCallback = this.filterCommandCallback;
            const filteredCommands = filterCallback ? yield array_1.filter(commands, (cmd) => tslib_1.__awaiter(this, void 0, void 0, function* () { return filterCallback(cmd); })) : commands;
            return array_1.map(filteredCommands, (cmd) => tslib_1.__awaiter(this, void 0, void 0, function* () { return this.formatCommand(cmd); }));
        });
    }
    formatCommand(cmd) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { command } = cmd;
            const formatter = new CommandSchemaHelpFormatter({
                location: { path: [...cmd.path], obj: command, args: [] },
                command,
                metadata: cmd,
            });
            return formatter.serialize();
        });
    }
}
exports.NamespaceSchemaHelpFormatter = NamespaceSchemaHelpFormatter;
class CommandSchemaHelpFormatter extends CommandHelpFormatter {
    format() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return JSON.stringify(yield this.serialize());
        });
    }
    serialize() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const metadata = yield this.getCommandMetadata();
            return this.formatCommand(metadata);
        });
    }
    formatInputs(inputs) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return Promise.all(inputs.map((input) => tslib_1.__awaiter(this, void 0, void 0, function* () { return this.formatInput(input); })));
        });
    }
    formatInput(input) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const name = input.name;
            const summary = input.summary;
            const required = input.validators && input.validators.includes(validators_1.validators.required) ? true : false;
            return { name, summary, required };
        });
    }
    formatOptions(options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const filterCallback = this.filterOptionCallback;
            const filteredOptions = filterCallback ? yield array_1.filter(options, (opt) => tslib_1.__awaiter(this, void 0, void 0, function* () { return filterCallback(opt); })) : options;
            return Promise.all(filteredOptions.map((opt) => tslib_1.__awaiter(this, void 0, void 0, function* () { return this.formatOption(opt); })));
        });
    }
    formatOption(option) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const name = option.name;
            const summary = option.summary ? option.summary.trim() : '';
            const groups = option.groups ? option.groups : [];
            const aliases = option.aliases ? option.aliases : [];
            const type = option.type ? option.type.name.toLowerCase() : 'string';
            return { name, type, summary, default: option.default, groups, aliases };
        });
    }
    formatCommand(cmd) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const commandPath = this.location.path.map(([p]) => p);
            const namespacePath = lodash.initial(commandPath);
            const name = commandPath.join(' ');
            const summary = cmd.summary ? cmd.summary.trim() : '';
            const description = cmd.description ? cmd.description.trim() : '';
            const groups = cmd.groups ? cmd.groups : [];
            const exampleCommands = cmd.exampleCommands ? cmd.exampleCommands.map(c => `${name} ${c}`) : [];
            const aliases = guards_1.isHydratedCommandMetadata(cmd) ? cmd.aliases : [];
            const inputs = cmd.inputs ? yield this.formatInputs(cmd.inputs) : [];
            const options = cmd.options ? yield this.formatOptions(cmd.options) : [];
            return { name, namespace: namespacePath, summary, description, groups, exampleCommands, aliases, inputs, options };
        });
    }
}
exports.CommandSchemaHelpFormatter = CommandSchemaHelpFormatter;
function formatOptionName(opt, { showAliases = true, colors = colors_1.DEFAULT_COLORS } = {}) {
    const { input } = colors;
    const showInverse = opt.type === Boolean && opt.default === true && opt.name.length > 1;
    return ((showInverse ? input(`--no-${opt.name}`) : input(`-${opt.name.length > 1 ? '-' : ''}${opt.name}`)) +
        (showAliases ?
            (!showInverse && opt.aliases && opt.aliases.length > 0 ? ', ' + opt.aliases
                .map(alias => input(`-${alias}`))
                .join(', ') : '') : ''));
}
exports.formatOptionName = formatOptionName;
function createCommandMetadataFromSchema(schema) {
    return {
        name: schema.name,
        summary: schema.summary,
        description: schema.description,
        groups: [...schema.groups],
        exampleCommands: [...schema.exampleCommands],
        inputs: [...schema.inputs],
        options: schema.options.map(opt => (Object.assign({}, opt, { type: opt.type === 'boolean' ? Boolean : String, groups: [...opt.groups], aliases: [...opt.aliases] }))),
    };
}
exports.createCommandMetadataFromSchema = createCommandMetadataFromSchema;
