// @ionic/utils-fs/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const fs = require("fs");
const os = require("os");
const path = require("path");
const stream = require("stream");
const through2 = require("through2");
const promisify = (func) => {
    return (...args) => {
        return new Promise((resolve, reject) => {
            func(...args, (err, response) => {
                if (err) {
                    return reject(err);
                }
                resolve(response);
            });
        });
    };
};
exports.access = promisify(fs.access);
exports.mkdir = promisify(fs.mkdir);
exports.open = promisify(fs.open);
exports.stat = promisify(fs.stat);
exports.unlink = promisify(fs.unlink);
exports.readFile = promisify(fs.readFile);
exports.writeFile = promisify(fs.writeFile);
exports.readDir = promisify(fs.readdir);
function statSafe(p) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        try {
            return yield exports.stat(p);
        }
        catch (e) {
            // ignore
        }
    });
}
exports.statSafe = statSafe;
function readDirp(dir, { filter, walkerOptions } = {}) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        return new Promise((resolve, reject) => {
            const items = [];
            let rs = walk(dir, walkerOptions);
            if (filter) {
                rs = rs.pipe(through2.obj(function (obj, enc, cb) {
                    if (!filter || filter(obj)) {
                        this.push(obj);
                    }
                    cb();
                }));
            }
            rs
                .on('error', (err) => reject(err))
                .on('data', (item) => items.push(item.path))
                .on('end', () => resolve(items));
        });
    });
}
exports.readDirp = readDirp;
function readDirSafe(dir) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        try {
            return yield exports.readDir(dir);
        }
        catch (e) {
            return [];
        }
    });
}
exports.readDirSafe = readDirSafe;
function readJsonFile(filePath, options = { encoding: 'utf8' }) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const f = yield exports.readFile(filePath, options);
        return JSON.parse(f);
    });
}
exports.readJsonFile = readJsonFile;
function writeJsonFile(filePath, json, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        return exports.writeFile(filePath, JSON.stringify(json, undefined, 2) + '\n', options);
    });
}
exports.writeJsonFile = writeJsonFile;
function fileToString(filePath) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        try {
            return yield exports.readFile(filePath, { encoding: 'utf8' });
        }
        catch (e) {
            if (e.code === 'ENOENT') {
                return '';
            }
            throw e;
        }
    });
}
exports.fileToString = fileToString;
function mkdirp(p, mode = 0o777) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const absPath = path.resolve(p);
        const pathObj = path.parse(absPath);
        const dirnames = absPath.split(path.sep).splice(1);
        const dirs = dirnames.map((v, i) => path.resolve(pathObj.root, ...dirnames.slice(0, i), v));
        for (const dir of dirs) {
            try {
                yield exports.mkdir(dir, mode);
            }
            catch (e) {
                if (e.code !== 'EEXIST') {
                    throw e;
                }
            }
        }
    });
}
exports.mkdirp = mkdirp;
function mkdirpSync(p, mode = 0o777) {
    const absPath = path.resolve(p);
    const pathObj = path.parse(absPath);
    const dirnames = absPath.split(path.sep).splice(1);
    const dirs = dirnames.map((v, i) => path.resolve(pathObj.root, ...dirnames.slice(0, i), v));
    for (const dir of dirs) {
        try {
            fs.mkdirSync(dir, mode);
        }
        catch (e) {
            if (e.code !== 'EEXIST') {
                throw e;
            }
        }
    }
}
exports.mkdirpSync = mkdirpSync;
function getFileChecksum(filePath) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const crypto = yield Promise.resolve().then(() => require('crypto'));
        return new Promise((resolve, reject) => {
            const hash = crypto.createHash('md5');
            const input = fs.createReadStream(filePath);
            input.on('error', (err) => {
                reject(err);
            });
            hash.once('readable', () => {
                const fullChecksum = hash.read().toString('hex');
                resolve(fullChecksum);
            });
            input.pipe(hash);
        });
    });
}
exports.getFileChecksum = getFileChecksum;
/**
 * Return true and cached checksums for a file by its path.
 *
 * Cached checksums are stored as `.md5` files next to the original file. If
 * the cache file is missing, the cached checksum is undefined.
 *
 * @param p The file path
 * @return Promise<[true checksum, cached checksum or undefined if cache file missing]>
 */
function getFileChecksums(p) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        return Promise.all([
            getFileChecksum(p),
            (() => tslib_1.__awaiter(this, void 0, void 0, function* () {
                try {
                    const md5 = yield exports.readFile(`${p}.md5`, { encoding: 'utf8' });
                    return md5.trim();
                }
                catch (e) {
                    if (e.code !== 'ENOENT') {
                        throw e;
                    }
                }
            }))(),
        ]);
    });
}
exports.getFileChecksums = getFileChecksums;
/**
 * Store a cache file containing the source file's md5 checksum hash.
 *
 * @param p The file path
 * @param checksum The checksum. If excluded, the checksum is computed
 */
function cacheFileChecksum(p, checksum) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const md5 = yield getFileChecksum(p);
        yield exports.writeFile(`${p}.md5`, md5, { encoding: 'utf8' });
    });
}
exports.cacheFileChecksum = cacheFileChecksum;
function writeStreamToFile(stream, destination) {
    return new Promise((resolve, reject) => {
        const dest = fs.createWriteStream(destination);
        stream.pipe(dest);
        dest.on('error', reject);
        dest.on('finish', resolve);
    });
}
exports.writeStreamToFile = writeStreamToFile;
function copyDirectory(source, destination, options = {}) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const ncp = yield Promise.resolve().then(() => require('ncp'));
        return new Promise((resolve, reject) => {
            ncp.ncp(source, destination, options, err => {
                if (err) {
                    reject(err);
                }
                resolve();
            });
        });
    });
}
exports.copyDirectory = copyDirectory;
function removeDirectory(dir) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const rimraf = yield Promise.resolve().then(() => require('rimraf'));
        const rimrafp = promisify(rimraf);
        yield rimrafp(dir);
    });
}
exports.removeDirectory = removeDirectory;
function copyFile(fileName, target, mode = 0o666) {
    return new Promise((resolve, reject) => {
        const rs = fs.createReadStream(fileName);
        const ws = fs.createWriteStream(target, { mode });
        rs.on('error', reject);
        ws.on('error', reject);
        ws.on('open', () => {
            rs.pipe(ws);
        });
        ws.once('finish', resolve);
    });
}
exports.copyFile = copyFile;
function pathAccessible(filePath, mode) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        try {
            yield exports.access(filePath, mode);
        }
        catch (e) {
            return false;
        }
        return true;
    });
}
exports.pathAccessible = pathAccessible;
function pathExists(filePath) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        return pathAccessible(filePath, fs.constants.F_OK);
    });
}
exports.pathExists = pathExists;
/**
 * Find the base directory based on the path given and a marker file to look for.
 */
function findBaseDirectory(dir, file) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        if (!dir || !file) {
            return;
        }
        for (const d of compilePaths(dir)) {
            const results = yield readDirSafe(d);
            if (results.includes(file)) {
                return d;
            }
        }
    });
}
exports.findBaseDirectory = findBaseDirectory;
/**
 * Generate a random file path within the computer's temporary directory.
 *
 * @param prefix Optionally provide a filename prefix.
 */
function tmpfilepath(prefix) {
    const rn = Math.random().toString(16).substring(2, 8);
    const p = path.resolve(os.tmpdir(), prefix ? `${prefix}-${rn}` : rn);
    return p;
}
exports.tmpfilepath = tmpfilepath;
/**
 * Given an absolute system path, compile an array of paths working backwards
 * one directory at a time, always ending in the root directory.
 *
 * For example, `'/some/dir'` => `['/some/dir', '/some', '/']`
 *
 * @param filePath Absolute system base path.
 */
function compilePaths(filePath) {
    filePath = path.normalize(filePath);
    if (!path.isAbsolute(filePath)) {
        throw new Error(`${filePath} is not an absolute path`);
    }
    const parsed = path.parse(filePath);
    if (filePath === parsed.root) {
        return [filePath];
    }
    return filePath
        .slice(parsed.root.length)
        .split(path.sep)
        .map((segment, i, array) => parsed.root + path.join(...array.slice(0, array.length - i)))
        .concat(parsed.root);
}
exports.compilePaths = compilePaths;
function isDir(p) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const stats = yield statSafe(p);
        if (stats && stats.isDirectory()) {
            return true;
        }
        return false;
    });
}
exports.isDir = isDir;
class Walker extends stream.Readable {
    constructor(p, options = {}) {
        super({ objectMode: true });
        this.p = p;
        this.options = options;
        this.paths = [this.p];
    }
    _read() {
        const p = this.paths.shift();
        const { pathFilter } = this.options;
        if (!p) {
            this.push(null); // tslint:disable-line:no-null-keyword
            return;
        }
        fs.stat(p, (err, stats) => {
            if (err) {
                this.emit('error', err);
                return;
            }
            const item = { path: p, stats };
            if (stats.isDirectory()) {
                fs.readdir(p, (err, contents) => {
                    if (err) {
                        this.emit('error', err);
                        return;
                    }
                    let paths = contents.map(file => path.join(p, file));
                    if (pathFilter) {
                        paths = paths.filter(p => pathFilter(p.substring(this.p.length + 1)));
                    }
                    this.paths.push(...paths);
                    this.push(item);
                });
            }
            else {
                this.push(item);
            }
        });
    }
}
exports.Walker = Walker;
function walk(p, options = {}) {
    return new Walker(p, options);
}
exports.walk = walk;
